### DISCUSSION TITLE
Feedback on shortcuts

### DISCUSSION THREAD

**Discussion by max-baz** [sayanarijit/xplr#146] [General]
I decided to take a closer look at built-in shortcuts and play with customizing them, thought I'd give you some feedback 🙂 

1. The "delete" shortcuts imho are unnecessarily noisy and complicated 🤔 

For `dd` I honestly simply see no usefulness, for me `rmdir` chokes on "directory not empty" all the time, I dont know when I'm supposed to use this shortcut, only if I pay attention to not select folders?...

`dD` is a bit noisy for my taste, removing a simple directory results in pages of output of all it's contents...

Are you using both shortcuts? I realize it's probably a matter of taste, but I'd look into making making output less noisy, and for myself I will map `dd` do `rmtrash -rf` and `dD` do `rm -rf`, so that both at least remove without choking what I tell them to remove, but one moves it to trash and another purges instantly.

2. I thought to remap `ctrl-a` to `a` with the following hack, it does work, but `a` does not appear in the help menu. Am I stepping in the unsupported area here? 😅  

```yaml
remaps:
  a: ctrl-a
  V: null
  ctrl-a: null
```

Same for `/`, I wanted to remove `ctrl-f` so that in help I only see `/`, but this seems to also hide the shortcut entirely from help:

```yaml
ctrl-f: null
```

It's probably because you have a notion of "primary shortcut" and "alias", but ideally as a user I'd not worry what is a main shortcut and what is an alias... but at least let's make it consistent, either prohibit me from doing this hack, or let's make sure the shortcut appears in the help.

3. For `open in gui`, I guess it all depends on user setup, so maybe don't change anything, but just FYI this seems to work perfectly well for me:

```yaml
v:
  help: open in gui
  messages:
    - BashExecSilently: |
        while IFS= read -r line; do xdg-open "${line:?}" >/dev/null 2>&1; done < "${XPLR_PIPE_RESULT_OUT:?}"
```

(recently I've been burned by the fact that `&>` was buggy and in some environments not 100% equal to `> .. 2>&1` so now I'm always using the latter one 😅 )

4. All in all I see a tendency in myself to remap keys to something more reachable, for example from day one I remapped `:e` to `e` (one less shortcut made me feel better), now I'm 100% remapping `gx` to `v` (thinking to avoid `x` in case my nnn muscle memory will confuse it with delete, so choosing between `p` and `v` for `p`re`v`iew), remapping `ctrl+a` to simply `a`, removing `ctrl+f` mapping since I have `/`. Create, copy and move I haven't really used yet, but it _feels_ like in my taste they belong to `c`, `y` and `p` in default mode.

5. This mapping I also added almost since day one, very useful for me, because arrow keys are too far off 😄 

```
search:
  key_bindings:
    remaps:
      tab: right
```

6. For next/prev visited path, still trying to get used to this feature, so not sure yet if I'll be searching for different shortcuts or unmapping them altogether 🤔 

7. I will eventually get to debugging logs, I dont know why yet, but `:l` flashes some screen which closes instantly, but at the same time `?` works correctly, so there must be some difference or bug somewhere...

Enough for now! 😁

**Comment by sayanarijit** [comment]
> Are you using both shortcuts? I realize it's probably a matter of taste, but I'd look into making making output less noisy, and for myself I will map dd do rmtrash -rf and dD do rm -rf, so that both at least remove without choking what I tell them to remove, but one moves it to trash and another purges instantly.

Unfortunately, I could not find any trashing utility that comes in-built in all the platforms and behaves the same way. If you go through the built-in commands, you will notice that I try to keep the use of "non-native" bash commands as little as possible. It can be argued that in platforms like `NetBSD`, even `bash` isn't "native", but in that case, bash will be the only dependency you will need to install (I don't want  my scripts to run in `sh`). The bottom line is that - I wasn't able to come up with a cross-platform, out of the box `rmtrash` alternative.

`dd` -> `rmdir` mapping for directory is simply to provide a consistent behavior, so users know what to expect when they run `dd` on a file or directory. Like "Deleting a single item? Use `dd`. Deleting recursively? Use `dD`". The script is written in a way that, when `dd` isn't able to delete nonempty directories, they are still selected so that you can then try `dD`. I mainly use `xplr` as `netrw` replacement from inside `neovim`, so the `rmdir` mapping sometimes come in handy when I create a directory and then want to delete it after realizing I could name it better. I don't have to get cautious I like I do to when running `dD`.

The verbosity is more of a personal preference. For me, I like watching what's happening when I'm executing some command. It would be nice if it supported progress bars,  but `-v` is all we get.

However, these mapping are meant to be overwritten based on your preference, platform, what tools you use and how you use `xplr`. The main reason these aren't implemented natively into `xplr` in rust.

When we implement [plugin support](https://github.com/sayanarijit/xplr/discussions/144) I think things will get better, and we'll be able to achieve more cross-platform way of doing things with more control.

>  thought to remap ctrl-a to a with the following hack, it does work, but a does not appear in the help menu. Am I stepping in the unsupported area here?

Yes :smile: "Un"mapping isn't supported as of now because remap works like symlink to existing mappings. In your example, the resolved mapping looks like `a -> ctrl-a -> null` i.e. `a` resolves to `null`. But I think we can "fix" this by using indexmap and "copying" the target action instead of "linking" to it. Same for `/`.

> For open in gui, I guess it all depends on user setup, so maybe don't change anything, but just FYI this seems to work perfectly well for me:

For me, I often open files while selecting to determine if I should select it or not. So I'd want `gx` to open the file under focus, not the selected ones.  Also, it might be useful to open multiple files in different tabs of a program like an editor, but I'm not sure when it makes sense to open multiple files in different windows.

> ecently I've been burned by the fact that &> was buggy and in some environments not 100% equal to > .. 2>&1 so now I'm always using the latter one

Good suggestion :+1:  I didn't know they're different.

> All in all I see a tendency in myself to remap keys to something more reachable, for example from day one I remapped :e to e (one less shortcut made me feel better), now I'm 100% remapping gx to v (thinking to avoid x in case my nnn muscle memory will confuse it with delete, so choosing between p and v for preview), remapping ctrl+a to simply a, removing ctrl+f mapping since I have /. Create, copy and move I haven't really used yet, but it feels like in my taste they belong to c, y and p in default mode.

The primary goal of the default key bindings is to demonstrate what's possible, and the goal of `xplr` config system it to help you customize it according to your preference with the least possible effort. I plan to include key bindings from `emacs`, `nnn`, `vifm`, `ranger`, etc. as [hacks](https://github.com/sayanarijit/xplr/wiki/Hacks) or [plugins](https://github.com/sayanarijit/xplr/discussions/144) once the API stabilizes, so that you can feel at home with `xplr` no matter where you come from.

> This mapping I also added almost since day one, very useful for me, because arrow keys are too far off

`tab` in search mode now maps to `select` to let you select files while searching (like fzf). I think `ctrl-n` and `ctrl-p` where [nice additions to the search mode](https://github.com/sayanarijit/xplr/pull/116) by @aflyingpumpkin that help you navigate vertically. I'm trying to figure out sensible horizontal navigation keys.

> For next/prev visited path, still trying to get used to this feature, so not sure yet if I'll be searching for different shortcuts or unmapping them altogether

I often use it to copy files to a destination and get back to the previous location. Also, when visiting a symlink and get back to the previous directory. But I guess we can make it "jump list" like in `vim` to make it more useful. It will help us navigate the history of jumps even in the same directory.

> I will eventually get to debugging logs, I dont know why yet, but :l flashes some screen which closes instantly, but at the same time ? works correctly, so there must be some difference or bug somewhere...

Weird! Try `cat $XPLR_PIPE_LOGS_OUT | ${PAGER:-less}` in the terminal.

**Reply by max-baz** [reply]
> Like "Deleting a single item? Use dd. Deleting recursively? Use dD".

Aah, this makes sense 🙂 And agree that people should override.

> but in that case, bash will be the only dependency you will need to install (I don't want my scripts to run in sh).

Maybe a little offtopic, I'm using kakoune editor (highly recommend btw), and it also supports writing hacks/plugins in shell. By default it uses bash, and when I migrated from neovim to kakoune and imported all the plugins, it was just super slow to start. It turned out that switching from bash to sh (dash) got super huge performance boost, it was completely game changing. 

In xplr not as many plugins will (probably?) run to observe such slowdown, but in any case it would be a good idea to be able to override if user wishes to.

In kakoune, it uses `/usr/bin/sh`, which on many systems is a symlink to bash, but people _can_ change the symlink to anything they like.

> For me, I often open files while selecting to determine if I should select it or not. So I'd want gx to open the file under focus, not the selected ones. 

Aah interesting - but that's not how `gx` works for me out of the box! 😄 

> tab in search mode now maps to select to let you select files while searching (like fzf). 

That's good to know, useful too!

> Weird! Try cat $XPLR_PIPE_LOGS_OUT | ${PAGER:-less} in the terminal.

Nice catch, it's of course me messing too much with less parameters 😅 I had `--quit-if-one-screen` in my `$LESS` 😁

**Reply by max-baz** [reply]
Regarding `less`, it actually is common than people override `$LESS` with various things. In my case this `$LESS` override even comes from zsh framework, and I kinda agree that it is usually a nice thing. 

Thus if xplr relies on the fact that it expects less to stay open, I think the best way to do it is to request this explicitly, i.e. pass `-+F` to less, like `cat file | less -+F`.

**Reply by sayanarijit** [reply]
I'm going with `LESS= less`

**Reply by max-baz** [reply]
This will be worse because for example I like to enable ignore-case search and I don't think xplr has to disable this... Just enable the feature you rely on, but don't disable things that users configured...

**Reply by sayanarijit** [reply]
Hmm... But with a custom LESS, there's no telling what other custom workarounds I might have to implement (like `-+F` for `--quit-if-one-screen`).

**Reply by sayanarijit** [reply]
Another option is maybe use the `PAGER` variable. This variable is meant to be interactive.

**Reply by max-baz** [reply]
Yeah I don't see one option that is definitely preferable... I as a user set `PAGER=less` and `LESS=someargs` and expect when pager is used by any app, to use less with my custom preferences. 

Ability to *not exit* is not a distinct feature of all pagers, so you can't just rely on the fact that pager will stay... If you knew that you'd be using less, it seems best to request this additional special feature with `-+F`, but given that you also allow custom `$PAGER` values, ones which potentially maybe dont have `-+F` flag... Dont know...

**Reply by sayanarijit** [reply]
I think `PAGER="less -+F"` is the correct fix. `$PAGER`, like `$EDITOR` is used in many programs that requires the user to review the contents of a file interactively before proceeding to the next step. Hence, setting `$PAGER` to something non-interactive could be fatal. As for `xplr`, adding the `-+F` won't really hurt, but the script won't be a one liner anymore :smile: .

**Reply by max-baz** [reply]
Can you think of a few examples of such programs where `$PAGER` is expected to be interactive? You are probably right, and I probably have not picked up on this trend yet, so would like to explore a bit 🙂

**Reply by sayanarijit** [reply]
Unfortunately I can't pinpoint to something right now, but I remember many times, specially during installation, or some similar multistep operation, the pager was invoked to display information between the steps.

**Reply by sayanarijit** [reply]
Oh, I think I can add the `-+F` by adding just one more line.

**Reply by sayanarijit** [reply]
> Aah interesting - but that's not how gx works for me out of the box! smile

WTF I don't remember when did I change the `gx` behavior :sweat_smile:  it used to do what I mentioned.

**Reply by sayanarijit** [reply]
Anyway, I'll leave it as it is. Preview probably should be a separate action.

**Reply by sayanarijit** [reply]
Nope. The `go to` mode actions handle only the cursor operations. multi-open should be `sx`.

**Reply by max-baz** [reply]
On a related note, is it possible for me to delete the file and preserve the cursor location, instead of it going to the top? A general case with multiple selection need some thought on how to do it properly, but in summary I'm imagining like `vim`, I am going through the directory, and what I dont like I just `dd` right away (I'm making my `dd` silent, not printing any output nor waiting for enter).

**Reply by sayanarijit** [reply]
Should be possible. But it doesn't work for some reason. Need to debug with some dedicated time.

**Reply by sayanarijit** [reply]
Basically 'FocusNext' needs to work after delete.

**Comment by max-baz** [comment]
Another related idea, that I'm not sure if is a good one or not: what if in the help menu we would sort the entries not by shortcut, but by description?

**Reply by sayanarijit** [reply]
I wanted to sort them by the definition order. i.e. use a `IndexMap`. But then I need to reorganize the existing keys with an order that makes sense. Seems lot of work:sweat_smile:  but what do think about this?

**Reply by max-baz** [reply]
Definition order would also need to figure out how to treat new user-defined keys and remaps (especially complicated ones, like `a->b, b->c, c->null` 🤔  Maybe don't go that road 😅 

The current order is helpful when you dont know what a key does (e.g. I know `d` does something, but I dont remember what, so thanks to sorted order by key I can quickly find it and read the description), but for me at least more often I dont know key combination for a function (e.g. I know there must be a "delete" feature, but I dont remember the key), that's why I thought sorting by descriptions could help with that...

**Reply by sayanarijit** [reply]
Ah, right... https://github.com/sayanarijit/xplr/issues/151

**Comment by max-baz** [comment]
Historically terminals couldn't distinguish certain shortcuts (e.g. `ctrl-i` and `tab`, or `ctrl+j` and `enter`), and although [this is being fixed now in modern terminals](https://sw.kovidgoyal.net/kitty/keyboard-protocol.html), I was surprised to discover that `xplr` already can distinguish `ctrl+j` and `enter` - but not `ctrl+i` and `tab`! How come? 

kakoune had to implement support for this new keyboard mode, I haven't looked deeply in xplr, but I presume you use some library for this? Would be real nice to track this, to be able to support wider range of shortcuts.

**Reply by sayanarijit** [reply]
I use [crossterm](https://github.com/crossterm-rs/crossterm). Not sure if `ctrl-i` and `tab` are distinguishable because kakoune and vim both seems to respond to `ctrl-i` and `tab` the same way.

**Reply by max-baz** [reply]
OK interesting, I'll open a thread there, lets see what they say. `ctri-i` and `tab` will be distinguishable in the future, it's just very new protocol, so in kakoune it's implemented in a branch, but not in latest release yet. I suspect neovim will have this support very soon too, if it doesnt have it already in some branch. 

FYI the way it works is that in "default" mode yes `ctrl-i` and `tab` produce the same character (you can check this by running `$ cat` and pressing `ctrl-i` and `tab`), but with the new protocol your app can send a special escape code, and after that terminal will be sending more info about pressed keys, so with that extra info you will see the difference.

**Reply by max-baz** [reply]
https://github.com/crossterm-rs/crossterm/issues/569

**Comment by max-baz** [comment]
Another idea to consider: remapping a key from one mode to another. For example I found myself remapping `e => :e` (to bring editor to default mode), or `gj => G` (to support kakoune's `gj` as go to the end). Right now I basically have to copy-paste the entire definition.

**Reply by sayanarijit** [reply]
Nice idea. `e: ":e"`.

**Reply by max-baz** [reply]
maybe we could denote a special character meaning "root", so if I'm in a custom mode A, I can bring a mapping from custom mode B using `m: "^bm"` (where `^` means "default mode")... ?

**Reply by max-baz** [reply]
or no, maybe simply if remap value contains more than one character, it already in itself means that it starts from default mode!

**Reply by sayanarijit** [reply]
~Then we'll lose the `^` key from remap options. Also~
Need to think more on this one. Because, even if we managed this with some syntax hack, It won't be very pretty looking.
There's another option to define actions separately and map keys to the definition of the actions by name. But it'll make things a little complicated.

**Reply by sayanarijit** [reply]
The single string idea won't work because there are keys like `esc`.

**Reply by max-baz** [reply]
True... It can only work if you can first parse them out into actual keypresses (like `esc` is one, `ctrl+shift+a` is another), and then count those...

**Reply by sayanarijit** [reply]
maybe having a global remap will have

**Reply by sayanarijit** [reply]
A `remaps` outside all the modes which maps a sequence of keys to another sequence of keys.

**Reply by sayanarijit** [reply]
Although it's possible to implement, I'm getting worried about the complexity config inheritance introduces. At this point I'd rather use a real embedded programming language than yaml.

**Reply by max-baz** [reply]
This could be interesting to explore! Any ideas for a specific language and examples / prototypes?

**Reply by sayanarijit** [reply]
So far 2 languages - `lua` and [gluon](https://github.com/gluon-lang/gluon). There's also [rhai](https://github.com/rhaiscript/rhai), but I'd prefer `gluon` if it has to be written in Rust. I'll create a prototype with both lua and gluon.

**Reply by sayanarijit** [reply]
Here's the `lua` binding: https://github.com/sayanarijit/xplr/pull/156.

**Reply by sayanarijit** [reply]
`gluon` seems to be the difficult one even though it's written in rust. I couldn't find any serde integration like in `lua`.

**Reply by sayanarijit** [reply]
WIP gluon experiment: https://github.com/sayanarijit/xplr/pull/157
